attribute-error on python 3.9

Reported via a user running the test-suite with py.test:

```
Traceback (most recent call last):
  File "/nix/store/917dzvfd7shvfzx827gp2gxlgd8ip49r-python3.9-Twisted-21.7.0/lib/python3.9/site-packages/twisted/internet/defer.py", line 858, in _runCallbacks
    current.result = callback(  # type: ignore[misc]
  File "/build/txtorcon-20.0.0/txtorcon/torinfo.py", line 220, in bootstrap
    d.addCallback(self._setup_complete)
  File "/nix/store/917dzvfd7shvfzx827gp2gxlgd8ip49r-python3.9-Twisted-21.7.0/lib/python3.9/site-packages/twisted/internet/defer.py", line 498, in addCallback
    return self.addCallbacks(callback, callbackArgs=args, callbackKeywords=kwargs)
  File "/nix/store/917dzvfd7shvfzx827gp2gxlgd8ip49r-python3.9-Twisted-21.7.0/lib/python3.9/site-packages/twisted/internet/defer.py", line 478, in addCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/nix/store/917dzvfd7shvfzx827gp2gxlgd8ip49r-python3.9-Twisted-21.7.0/lib/python3.9/site-packages/twisted/internet/defer.py", line 858, in _runCallbacks
    current.result = callback(  # type: ignore[misc]
  File "/build/txtorcon-20.0.0/txtorcon/torinfo.py", line 285, in _setup_complete
    pb.callback(self)
  File "/nix/store/917dzvfd7shvfzx827gp2gxlgd8ip49r-python3.9-Twisted-21.7.0/lib/python3.9/site-packages/twisted/internet/defer.py", line 661, in callback
    assert not isinstance(result, Deferred)
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
  File "/build/txtorcon-20.0.0/txtorcon/torinfo.py", line 213, in __getattribute__
    raise AttributeError(name)
builtins.AttributeError: __class__

self = <test.test_web.WebAgentTests testMethod=test_agent>
result = <TestCaseFunction test_agent>

    def run(self, result):
        """
        Run the test case, storing the results in C{result}.

        First runs C{setUp} on self, then runs the test method (defined in the
        constructor), then runs C{tearDown}.  As with the standard library
        L{unittest.TestCase}, the return value of these methods is disregarded.
        In particular, returning a L{Deferred<twisted.internet.defer.Deferred>}
        has no special additional consequences.

        @param result: A L{TestResult} object.
        """
        log.msg("--> %s <--" % (self.id()))
        new_result = itrial.IReporter(result, None)
        if new_result is None:
            result = PyUnitResultAdapter(result)
        else:
            result = new_result
        result.startTest(self)
        (doSkip, skipReason) = self.getSkip()
        if doSkip:  # don't run test methods that are marked as .skip
>           result.addSkip(self, skipReason)
E           TypeError: Skipped expected string as 'msg' parameter, got 'bool' instead.
E           Perhaps you meant to use a mark?
```

--- a/.github/workflows/flake8.yaml
+++ b/.github/workflows/flake8.yaml
@@ -1,0 +1,28 @@
+name: flake8
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  testing:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.9
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip tox tox-gh-actions
+
+    - name: Test
+      run: |
+        python --version
+        python -m tox -e flake8

--- a/.github/workflows/python2.yaml
+++ b/.github/workflows/python2.yaml
@@ -1,0 +1,34 @@
+name: python2
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  testing:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [2.7]
+## XXX not sure how to tie this to "tox -e tx21" below, for example (just get GHA errors)
+##        twisted-version: [tx17, tx18, tx19, tx20, tx21]
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip tox tox-gh-actions
+
+    - name: Test
+      run: |
+        python --version
+        python -m tox
+##        python -m tox -e ${{ matrix.twisted-version }}

--- a/.github/workflows/python2.yaml
+++ b/.github/workflows/python2.yaml
@@ -12,8 +12,7 @@ jobs:
     strategy:
       matrix:
         python-version: [2.7]
-## XXX not sure how to tie this to "tox -e tx21" below, for example (just get GHA errors)
-##        twisted-version: [tx17, tx18, tx19, tx20, tx21]
+        twisted-version: [tx17, tx18, tx19, tx20, tx21]
 
     steps:
     - uses: actions/checkout@v2
@@ -27,8 +26,9 @@ jobs:
       run: |
         python -m pip install --upgrade pip tox tox-gh-actions
 
-    - name: Test
+    - name: Unit-Test
+      env:
+        TOX_ENV: ${{ matrix.twisted-version }}
       run: |
         python --version
         python -m tox
-##        python -m tox -e ${{ matrix.twisted-version }}

--- a/.github/workflows/python3.yaml
+++ b/.github/workflows/python3.yaml
@@ -12,8 +12,7 @@ jobs:
     strategy:
       matrix:
         python-version: [3.5, 3.6, 3.7, 3.8, 3.9]
-## XXX not sure how to tie this to "tox -e tx21" below, for example (just get GHA errors)
-##        twisted-version: [tx17, tx18, tx19, tx20, tx21]
+        twisted-version: [tx17, tx18, tx19, tx20, tx21]
 
     steps:
     - uses: actions/checkout@v2
@@ -28,7 +27,8 @@ jobs:
         python -m pip install --upgrade pip tox tox-gh-actions
 
     - name: Test
+      env:
+        TOX_ENV: ${{ matrix.twisted-version }}
       run: |
         python --version
         python -m tox
-##        python -m tox -e ${{ matrix.twisted-version }}

--- a/.github/workflows/python3.yaml
+++ b/.github/workflows/python3.yaml
@@ -1,0 +1,31 @@
+name: python3
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  testing:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [3.5, 3.6, 3.7, 3.8, 3.9]
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip tox tox-gh-actions
+
+    - name: Test
+      run: |
+        python --version
+        tox

--- a/.github/workflows/python3.yaml
+++ b/.github/workflows/python3.yaml
@@ -29,4 +29,4 @@ jobs:
     - name: Test
       run: |
         python --version
-        python -m tox -e ${{ matrix.twisted_version }}
+        python -m tox -e ${{ matrix.twisted-version }}

--- a/.github/workflows/python3.yaml
+++ b/.github/workflows/python3.yaml
@@ -28,4 +28,4 @@ jobs:
     - name: Test
       run: |
         python --version
-        tox
+        python -m tox

--- a/.github/workflows/python3.yaml
+++ b/.github/workflows/python3.yaml
@@ -12,7 +12,8 @@ jobs:
     strategy:
       matrix:
         python-version: [3.5, 3.6, 3.7, 3.8, 3.9]
-        twisted-version: [tx17, tx18, tx19, tx20, tx21]
+## XXX not sure how to tie this to "tox -e tx21" below, for example (just get GHA errors)
+##        twisted-version: [tx17, tx18, tx19, tx20, tx21]
 
     steps:
     - uses: actions/checkout@v2
@@ -29,4 +30,5 @@ jobs:
     - name: Test
       run: |
         python --version
-        python -m tox -e ${{ matrix.twisted-version }}
+        python -m tox
+##        python -m tox -e ${{ matrix.twisted-version }}

--- a/.github/workflows/python3.yaml
+++ b/.github/workflows/python3.yaml
@@ -12,7 +12,7 @@ jobs:
     strategy:
       matrix:
         python-version: [3.5, 3.6, 3.7, 3.8, 3.9]
-        twisted-version: [17, 18, 19, 20, 21]
+        twisted-version: [tx17, tx18, tx19, tx20, tx21]
 
     steps:
     - uses: actions/checkout@v2
@@ -29,4 +29,4 @@ jobs:
     - name: Test
       run: |
         python --version
-        python -m tox -e tx${{ matrix.twisted_version }}
+        python -m tox -e ${{ matrix.twisted_version }}

--- a/.github/workflows/python3.yaml
+++ b/.github/workflows/python3.yaml
@@ -26,7 +26,7 @@ jobs:
       run: |
         python -m pip install --upgrade pip tox tox-gh-actions
 
-    - name: Test
+    - name: Unit-Test
       env:
         TOX_ENV: ${{ matrix.twisted-version }}
       run: |

--- a/.github/workflows/python3.yaml
+++ b/.github/workflows/python3.yaml
@@ -12,6 +12,7 @@ jobs:
     strategy:
       matrix:
         python-version: [3.5, 3.6, 3.7, 3.8, 3.9]
+        twisted-version: [17, 18, 19, 20, 21]
 
     steps:
     - uses: actions/checkout@v2
@@ -28,4 +29,4 @@ jobs:
     - name: Test
       run: |
         python --version
-        python -m tox
+        python -m tox -e tx${{ matrix.twisted_version }}

--- a/.github/workflows/readme_render.yaml
+++ b/.github/workflows/readme_render.yaml
@@ -1,0 +1,28 @@
+name: README render
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  testing:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.9
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip tox tox-gh-actions
+
+    - name: Test
+      run: |
+        python --version
+        python -m tox -e readme_render

--- a/docs/releases.rst
+++ b/docs/releases.rst
@@ -15,7 +15,10 @@ that will be "19.2.1".
 
 See also :ref:`api_stability`.
 
-`git main <https://github.com/meejah/txtorcon>`_ *will likely become v20.1.0*
+`git main <https://github.com/meejah/txtorcon>`_ *will likely become v21.0.0*
+
+ * Use GitHub Actions instead of Travis for CI (needs more work)
+ * fix Python 3.x+ bug with `TorInfo` and `__class__` access (`#350 <https://github.com/meejah/txtorcon/issues/350>`_)
 
 
 v20.0.0

--- a/test/test_util.py
+++ b/test/test_util.py
@@ -3,7 +3,7 @@ import sys
 import tempfile
 import ipaddress
 from mock import patch
-from unittest import skipIf
+from unittest import skip
 from os.path import exists
 
 from twisted.trial import unittest
@@ -62,7 +62,7 @@ class TestGeoIpDatabaseLoading(unittest.TestCase):
         util.GeoIP = _GeoIP
         self.assertEqual(ret_val, None)
 
-    @skipIf('pypy' in sys.version.lower(), "No GeoIP in PyPy")
+    @skip("No GeoIP in github-actions")
     def test_return_geoip_object(self):
         # requires a valid GeoIP database to work, so hopefully we're
         # on Debian or similar...

--- a/test/test_util.py
+++ b/test/test_util.py
@@ -1,5 +1,4 @@
 import os
-import sys
 import tempfile
 import ipaddress
 from mock import patch

--- a/tox.ini
+++ b/tox.ini
@@ -39,6 +39,14 @@ deps=
     tx20: twisted>=20.0.0,<21.0.0
     tx21: twisted>=21.0.0,<22.0.0
 
+[gh-actions]
+python =
+  3.5: py35
+  3.6: py36
+  3.7: py37
+  3.8: py38
+  3.9: py39
+
 [testenv:clean]
 deps=
     coverage

--- a/tox.ini
+++ b/tox.ini
@@ -11,7 +11,7 @@
 # stretch: 16.3.0-1
 
 [tox]
-envlist = clean,flake8,stats,readme_render,{py27,py36,py37,py38,pypy}-{tx16,tx17,tx18,tx19,tx20}
+envlist = clean,flake8,stats,readme_render,{py27,py36,py37,py38,py39,pypy}-{tx16,tx17,tx18,tx19,tx20,tx21}
 # if you're not using detox, you can use this list to get
 # "all environments" coverage stats:
 # tox -e 'clean,flake8,py27-tx18,pypy-tx18,py35-tx18,stats'
@@ -36,6 +36,8 @@ deps=
     tx17: twisted>=17.0.0,<18.0.0
     tx18: twisted>=18.0.0,<19.0.0
     tx19: twisted>=19.0.0,<20.0.0
+    tx20: twisted>=20.0.0,<21.0.0
+    tx21: twisted>=21.0.0,<22.0.0
 
 [testenv:clean]
 deps=

--- a/tox.ini
+++ b/tox.ini
@@ -29,7 +29,8 @@ deps=
     setuptools>=0.8.0
     automat
     mock
-    GeoIP
+## XXX this doesn't work on github-actions
+##    GeoIP
     coverage
     cryptography
     tx16: twisted>=16.0.0,<17.0.0

--- a/txtorcon/torinfo.py
+++ b/txtorcon/torinfo.py
@@ -199,8 +199,11 @@ class TorInfo(object):
             return sup.__getattribute__(name)
 
         attrs = sup.__getattribute__('attrs')
+        # are there other "special" attributes we need to consider..?
         if name == '__members__':
             return list(attrs.keys())
+        if name == '__class__':
+            return sup.__class__
 
         else:
             try:

